### PR TITLE
feat(auth): AppKit Auth

### PR DESCRIPTION
**Pull request by Cali93**
### Context

Introduces [AppKit Auth](https://www.notion.so/walletconnect/AppKit-Auth-13dcc2ef3ba14b698bb1eeea87b7442f).
Notable changes:
- `chains` has been removed from `getMessageParams`, we're relying on the chains provided to the connector
- By using `enableAuth: true` developers don't need to provide a `siweConfig`. A default config is used instead.

## Type of change

- [ ] Chore (non-breaking change that addresses non-functional tasks, maintenance, or code quality improvements)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# Associated Issues

For Linear issues: Closes APKT-xxx
For GH issues: closes #...

# Showcase (Optional)

If there is a UI change include the screenshots with before and after state.
If new feature is being introduced, include the link to demo recording.

# Checklist

- [ ] Code in this PR is covered by automated tests (Unit tests, E2E tests)
- [ ] My changes generate no new warnings
- [ ] I have reviewed my own code
- [ ] I have filled out all required sections
- [ ] I have tested my changes on the preview link
- [ ] Approver of this PR confirms that the changes are tested on the preview link

### CODE REVIEW

File: apps/laboratory/src/pages/library/wagmi-siwe.tsx [hunk 1]
```diff
@@ -19,6 +19,7 @@ const modal = createWeb3Modal({
   enableAnalytics: true,
   metadata: ConstantsUtil.Metadata,
   siweConfig,
+  enableOnramp: true,
```
Comment (enesozturk) [inline apps/laboratory/src/pages/library/wagmi-siwe.tsx:22]: Is this onramp flag related with this PR? This should be true by default that we recently did this. Could you double check?

File: packages/common/src/utils/ConstantsUtil.ts [hunk 2]
```diff
@@ -2,5 +2,6 @@ export const ConstantsUtil = {
   WC_NAME_SUFFIX: '.wcn.id',
   BLOCKCHAIN_API_RPC_URL: 'https://rpc.walletconnect.org',
   PULSE_API_URL: 'https://pulse.walletconnect.org',
-  W3M_API_URL: 'https://api.web3modal.org'
+  W3M_API_URL: 'https://api.web3modal.org',
+  APPKIT_AUTH_API_URL: 'https://api-web3modal-staging.walletconnect-v1-bridge.workers.dev'
```
Comment (enesozturk) [inline packages/common/src/utils/ConstantsUtil.ts:6]: Is this expected to be changed? Do we some changes to be deployed to w3m API?

File: packages/core/src/controllers/ApiController.ts [hunk 3]
```diff
@@ -230,17 +230,16 @@ export const ApiController = {
...
-  async fetchAnalyticsConfig() {
-    const { isAnalyticsEnabled } = await api.get<ApiGetAnalyticsConfigResponse>({
+  async fetchProjectConfig() {
+    const { isAnalyticsEnabled } = await api.get<ApiGetProjectConfigResponse>({
       path: '/getAnalyticsConfig',
       headers: ApiController._getApiHeaders()
     })
-    OptionsController.setEnableAnalytics(isAnalyticsEnabled)
+
+    return { isAnalyticsEnabled, isAppKitAuthEnabled: true }
```
Comment (enesozturk) [inline packages/core/src/controllers/ApiController.ts:243]: If it's already hardcoded as `true`, why do we need to return this value from this function? Will it be read from the response of any endpoint later?

Comment (enesozturk) [reply packages/core/src/controllers/ApiController.ts:243]: Also the project name changed but it still uses `/getAnalyticsConfig`, is it expected?

File: packages/core/src/controllers/ApiController.ts [hunk 4]
```diff
@@ -230,17 +230,16 @@ export const ApiController = {
...
-    }
+
     state.prefetchPromise = Promise.race([Promise.allSettled(promises), CoreHelperUtil.wait(3000)])
   },
 
-  async fetchAnalyticsConfig() {
-    const { isAnalyticsEnabled } = await api.get<ApiGetAnalyticsConfigResponse>({
+  async fetchProjectConfig() {
+    const { isAnalyticsEnabled } = await api.get<ApiGetProjectConfigResponse>({
       path: '/getAnalyticsConfig',
```
Comment (enesozturk) [inline packages/core/src/controllers/ApiController.ts:239]: As I see the `/getProjectConfig` is called in tests but here we still use `/getAnalyticsConfig`. @Cali93

File: packages/scaffold/src/client.ts [hunk 5]
```diff
@@ -309,10 +310,17 @@ export class Web3ModalScaffold {
       NetworkController.setAllowUnsupportedChain(options.allowUnsupportedChain)
     }
 
-    if (options.siweControllerClient) {
-      const { SIWEController } = await import('@web3modal/siwe')
+    const { isAppKitAuthEnabled, isAnalyticsEnabled } = await ApiController.fetchProjectConfig()
```
Comment (enesozturk) [inline packages/scaffold/src/client.ts:313]: What about separating the `isAppKitAuthEnabled` and `isAnalyticsEnabled` into two different functions and make the fetch if it's necessary. 

I see there is a case that code might not go into both if's and in this case this async call will be unnecessary.

File: packages/siwe/core/utils/AppKitAuthUtil.ts [hunk 6]
```diff
@@ -0,0 +1,108 @@
+import { formatMessage } from '@walletconnect/utils'
+import { Web3ModalSIWEClient } from '../../src/client.js'
+import type { SIWECreateMessageArgs, SIWESession, SIWEVerifyMessageArgs } from './TypeUtils.js'
+import { ConstantsUtil } from '@web3modal/common'
+
+const myHeaders = new Headers()
+myHeaders.append('x-project-id', '24970167f11c121f6eb40b558edb9691')
```
Comment (enesozturk) [inline packages/siwe/core/utils/AppKitAuthUtil.ts:7]: I guess this is the problem where you wanted to get these values dynamically right?

File: apps/laboratory/src/pages/library/wagmi-siwe.tsx (hunk 1 above)
Comment (Cali93) [reply apps/laboratory/src/pages/library/wagmi-siwe.tsx:22]: Not related not sure how it ended up here. I'll double check 👍

File: packages/common/src/utils/ConstantsUtil.ts (hunk 2 above)
Comment (Cali93) [reply packages/common/src/utils/ConstantsUtil.ts:6]: No this is temporary until the Auth API routes are live in prod

File: packages/core/src/controllers/ApiController.ts (hunk 3 above)
Comment (Cali93) [reply packages/core/src/controllers/ApiController.ts:243]: Yes and yes it's temporary until `/getProjectConfig` is live on prod.

File: packages/scaffold/src/client.ts (hunk 5 above)
Comment (Cali93) [reply packages/scaffold/src/client.ts:313]: We always have to fetch `isAppKitAuthEnabled` from now.

File: packages/siwe/core/utils/AppKitAuthUtil.ts (hunk 6 above)
Comment (Cali93) [reply packages/siwe/core/utils/AppKitAuthUtil.ts:7]: No this is temporary until the API auth routes are live in prod.
This is a Cloud staging project id.

File: packages/core/src/controllers/ApiController.ts (hunk 4 above)
Comment (Cali93) [reply packages/core/src/controllers/ApiController.ts:239]: It'll be renamed to `getProjectConfig` when the routes are live on prod

**Comment by bkrem** [comment]
Couple of things that came up while testing this so far (see attached video):
1. `/me` seems to be doing a duplicate call on `Connect` with identical payload returns
2.  I got myself into some bad state where `/authenticate` always gives me 404 but no indication of what's wrong (repro'd both with Rainbow and Zerion)

Repro for 2.:
1. I connect first time, get all the way through with /authenticate 200
2. Disconnect -> observe `/sign-out`
3. Restart the flow -> get 404 on `/authenticate`

https://github.com/WalletConnect/web3modal/assets/8663099/c75666ea-6182-480c-b6d7-f99340a130df

File: packages/adapters/wagmi/src/client.ts [hunk 7]
```diff
@@ -282,49 +282,70 @@ export class WagmiAdapter implements ChainAdapter {
...
 
-        const siweParams = await this.options?.siweConfig?.getMessageParams?.()
+        const clientId = await provider.client?.core?.crypto?.getClientId()
+        if (clientId) {
+          this.appKit?.setClientId(clientId)
+        }
+
+        let chainId = Number(NetworkUtil.caipNetworkIdToNumber(this.appKit?.getCaipNetwork()?.id))
+        let address: string | undefined = undefined
+        let isSuccessful1CA = false
```
Comment (magiziz) [inline packages/adapters/wagmi/src/client.ts:292]: Can we do `isSuccessfulOneCa` ?

File: packages/adapters/wagmi/src/client.ts [hunk 8]
```diff
@@ -282,49 +282,70 @@ export class WagmiAdapter implements ChainAdapter {
...
-        const isSiweEnabled = this.options?.siweConfig?.options?.enabled
+        const isSiweEnabled = this.appKit?.getIsSiweEnabled()
         const isProviderSupported = typeof provider?.authenticate === 'function'
-        const isSiweParamsValid = siweParams && Object.keys(siweParams || {}).length > 0
-        const siweConfig = this.options?.siweConfig
-
-        if (isSiweEnabled && isProviderSupported && isSiweParamsValid && siweConfig) {
-          // @ts-expect-error - setting requested chains beforehand avoids wagmi auto disconnecting the session when `connect` is called because it things chains are stale
-          await connector.setRequestedChainsIds(siweParams.chains)
+        const supports1ClickAuth = isSiweEnabled && isProviderSupported
```
Comment (magiziz) [inline packages/adapters/wagmi/src/client.ts:296]: Sam here `supportsOneClickAuth` ?

**Review by magiziz — APPROVED** [review]
lgtm

File: apps/laboratory/src/utils/ProfilesUtil.ts [hunk 9]
```diff
@@ -0,0 +1,235 @@
+import {
```
Comment (tomiir) [inline apps/laboratory/src/utils/ProfilesUtil.ts:1]: are these utils intended to be implemented by developers?

File: apps/laboratory/src/utils/ProfilesUtil.ts [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+  type SIWECreateMessageArgs,
+  type SIWESession,
+  type SIWEVerifyMessageArgs
+} from '@reown/appkit-siwe'
+
+import { type Config, disconnect } from '@wagmi/core'
+import { ProfileStore } from './ProfileStoreUtil'
+const queryParams = `projectId=24970167f11c121f6eb40b558edb9691&st=w3m&sv=5.0.0`
+
+const profilesApiUrl = 'https://api-web3modal-staging.walletconnect-v1-bridge.workers.dev'
```
Comment (tomiir) [inline apps/laboratory/src/utils/ProfilesUtil.ts:17]: remove before merge?

File: packages/adapters/wagmi/src/client.ts [hunk 11]
```diff
@@ -335,22 +356,36 @@ export class WagmiAdapter implements ChainAdapter {
...
                 message,
                 signature: s.s,
-                cacao: signedCacao
+                cacao: signedCacao,
+                clientId
               })
+              isSuccessful1CA = true
             } catch (error) {
+              isSuccessful1CA = false
+              SIWEController.setIs1ClickAuthenticating(false)
```
Comment (tomiir) [inline packages/adapters/wagmi/src/client.ts:365]: how are these flags different?

File: packages/appkit/src/client.ts [hunk 12]
```diff
@@ -470,11 +474,34 @@ export class AppKit {
       adapter => adapter.chainNamespace === ConstantsUtil.CHAIN.EVM
     )
 
+    // Only set the analytics state if it's not already set through the SDK config
+    if (options.features?.analytics === undefined) {
+      const projectCloudConfig = await ApiController.fetchProjectConfig()
+      OptionsController.setFeatures({ analytics: projectCloudConfig?.isAppKitAuthEnabled })
+
+      if (options.enableAuth === undefined) {
```
Comment (tomiir) [inline packages/appkit/src/client.ts:482]: can we use default params instead of checking against `undefined` ?

File: packages/common/src/utils/TypeUtil.ts [hunk 13]
```diff
@@ -111,7 +111,14 @@ type BalanceQuantity = {
...
 
-export type SIWEStatus = 'uninitialized' | 'ready' | 'loading' | 'success' | 'rejected' | 'error'
+export type SIWEStatus =
+  | 'uninitialized'
+  | 'ready'
+  | 'loading'
+  | 'success'
+  | 'rejected'
+  | 'error'
+  | 'authenticating'
```
Comment (tomiir) [inline packages/common/src/utils/TypeUtil.ts:121]: how is it different than loading?

File: packages/core/src/controllers/AccountController.ts [hunk 14]
```diff
@@ -37,9 +36,17 @@ export interface AccountControllerState {
   preferredAccountType?: W3mFrameTypes.AccountType
   socialWindow?: Window
   farcasterUrl?: string
+  is1ClickAuthenticating?: boolean
   provider?: UniversalProvider | Provider | CombinedProvider
   status?: 'reconnecting' | 'connected' | 'disconnected' | 'connecting'
-  siweStatus?: 'uninitialized' | 'ready' | 'loading' | 'success' | 'rejected' | 'error'
+  siweStatus?:
```
Comment (tomiir) [inline packages/core/src/controllers/AccountController.ts:42]: can't you use `SiweStatus` ?

File: apps/laboratory/src/utils/ProfilesUtil.ts (hunk 9 above)
Comment (Cali93) [reply apps/laboratory/src/utils/ProfilesUtil.ts:1]: Nope it was for testing purposes around profiles.

File: apps/laboratory/src/utils/ProfilesUtil.ts (hunk 10 above)
Comment (Cali93) [reply apps/laboratory/src/utils/ProfilesUtil.ts:17]: Yes it'll use the prod API before merge

File: packages/scaffold-ui/src/views/w3m-connecting-wc-view/index.ts [hunk 15]
```diff
@@ -28,11 +29,30 @@ export class W3mConnectingWcView extends LitElement {
...
+        if (val.siweStatus === 'authenticating') {
+          SnackController.showLoading('Authenticating', 8000)
+        }
+
+        if (val.siweStatus === 'success') {
+          SnackController.hide()
+        }
+        if (val.siweStatus === 'ready') {
+          SnackController.hide()
+        }
```
Comment (tomiir) [inline packages/scaffold-ui/src/views/w3m-connecting-wc-view/index.ts:52]: what about other states? do they need handling?

File: packages/adapters/wagmi/src/client.ts (hunk 11 above)
Comment (Cali93) [reply packages/adapters/wagmi/src/client.ts:365]: `isSuccessful1CA` is to know if the 1CA flow was successful and run the `onSignIn` logic if provided in the `siweConfig` and it's only used here.
`setIs1ClickAuthenticating` is used to show a loader on the QR code screen and it's in the global state

File: packages/common/src/utils/TypeUtil.ts (hunk 13 above)
Comment (Cali93) [reply packages/common/src/utils/TypeUtil.ts:121]: authenticating is specific to 1ClickAuth

File: packages/core/src/controllers/AccountController.ts (hunk 14 above)
Comment (Cali93) [reply packages/core/src/controllers/AccountController.ts:42]: Nope because you'd have to import the SIWE package

File: packages/adapters/wagmi/src/client.ts (hunk 7 above)
Comment (Cali93) [reply packages/adapters/wagmi/src/client.ts:292]: 1CA is the naming convention we've been using, I'd rather stick with it.

File: packages/appkit/src/client.ts (hunk 12 above)
Comment (Cali93) [reply packages/appkit/src/client.ts:482]: So we're just removing the dependency on getting the config from Cloud ?
The initial product requirement was, if `enableAnalytics` is not set on the AppKit config, then check the Cloud config. By introducing a default value, you're breaking that flow.

File: packages/scaffold-ui/src/views/w3m-connecting-wc-view/index.ts (hunk 15 above)
Comment (Cali93) [reply packages/scaffold-ui/src/views/w3m-connecting-wc-view/index.ts:52]: Other states don't need to be handled.

File: packages/appkit/src/client.ts (hunk 12 above)
Comment (tomiir) [reply packages/appkit/src/client.ts:499]: I see, a bit of a weird flow imo but if that was the product decision that's ok

File: packages/siwe/package.json [hunk 16]
```diff
@@ -25,6 +25,24 @@
...
+    "react": "18.2.0",
+    "react-dom": "18.2.0"
+  },
+  "peerDependenciesMeta": {
+    "react": {
+      "optional": true
+    },
+    "react-dom": {
+      "optional": true
+    }
```
Comment (enesozturk) [inline packages/siwe/package.json:44]: Why react deps have been added here?

File: packages/scaffold-ui/exports/index.ts [hunk 17]
```diff
@@ -10,109 +10,96 @@ export * from '../src/modal/w3m-account-button/index.js'
 export * from '../src/modal/w3m-button/index.js'
```
Comment (enesozturk) [inline packages/scaffold-ui/exports/index.ts:10]: Revert these changes since they are unnecessary

File: packages/appkit/src/client.ts [hunk 18]
```diff
@@ -1,45 +1,45 @@
+import {
+  ConstantsUtil,
+  SafeLocalStorage,
+  SafeLocalStorageKeys,
+  type CaipNetwork,
+  type ChainNamespace
+} from '@reown/appkit-common'
 import type {
```
Comment (enesozturk) [inline packages/appkit/src/client.ts:8]: Revert import changes since they are unnecessary

File: packages/scaffold-ui/exports/index.ts (hunk 17 above)
Comment (enesozturk) [reply packages/scaffold-ui/exports/index.ts:10]: New changes have missing exported screens and breaking the app

**Comment by tomiir** [comment]
Superseded by Cloud Auth development
